File: backend-rust/src/services/google_index_queue.rs
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

use serde::{Deserialize, Serialize};
use serde_json::{json, Map, Value};
use tracing::{info, warn};

const QUEUE_FILE: &str = "google-index-queue.json";
const HISTORY_FILE: &str = "google-index-history.json";
const MAX_RETRY_COUNT: i32 = 3;
const DEDUP_DAYS: i64 = 30;
const MAX_HISTORY_ENTRIES_PER_URL: usize = 10;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UrlSubmissionQueue {
    pub url: String,
    pub action: String,
    #[serde(default = "default_priority")]
    pub priority: i32,
    #[serde(default = "now_ms")]
    pub queued_at: i64,
    #[serde(default)]
    pub retry_count: i32,
    #[serde(default)]
    pub last_error: Option<String>,
    #[serde(default = "default_status")]
    pub status: String,
    #[serde(default)]
    pub last_attempt_at: i64,
}

fn default_priority() -> i32 {
    5
}
fn default_status() -> String {
    "PENDING".into()
}
fn now_ms() -> i64 {
    chrono::Utc::now().timestamp_millis()
}

impl UrlSubmissionQueue {
    pub fn new(url: impl Into<String>, action: impl Into<String>, priority: i32) -> Self {
        Self {
            url: url.into(),
            action: action.into(),
            priority: priority.clamp(1, 10),
            queued_at: now_ms(),
            retry_count: 0,
            last_error: None,
            status: "PENDING".into(),
            last_attempt_at: 0,
        }
    }

    pub fn effective_priority(&self) -> f64 {
        let age_hours = (now_ms() - self.queued_at) as f64 / (1000.0 * 60.0 * 60.0);
        (11 - self.priority) as f64 * 100.0 + age_hours
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UrlSubmissionHistory {
    pub url: String,
    pub action: String,
    #[serde(default = "now_ms")]
    pub submitted_at: i64,
    #[serde(default)]
    pub submitted_date: String,
    pub success: bool,
    #[serde(default)]
    pub response: Option<String>,
    #[serde(default)]
    pub quota_used: i32,
    #[serde(default = "gen_submission_id")]
    pub submission_id: String,
}

fn gen_submission_id() -> String {
    format!("{}-{}", now_ms(), std::time::SystemTime::now().elapsed().unwrap_or_default().as_nanos())
}

impl UrlSubmissionHistory {
    pub fn new(url: impl Into<String>, action: impl Into<String>, success: bool) -> Self {
        let ts = now_ms();
        Self {
            url: url.into(),
            action: action.into(),
            submitted_at: ts,
            submitted_date: chrono::Local::now().format("%Y-%m-%d").to_string(),
            success,
            response: None,
            quota_used: 0,
            submission_id: gen_submission_id(),
        }
    }

    pub fn is_within_days(&self, days: i64) -> bool {
        now_ms() - self.submitted_at < days * 24 * 60 * 60 * 1000
    }
}

pub struct GoogleIndexQueueService {
    queue_file: PathBuf,
    history_file: PathBuf,
    inner: Mutex<QueueInner>,
}

struct QueueInner {
    queue: HashMap<String, UrlSubmissionQueue>,
    history: HashMap<String, Vec<UrlSubmissionHistory>>,
}

impl GoogleIndexQueueService {
    pub fn new(work_dir: PathBuf) -> Self {
        let svc = Self {
            queue_file: work_dir.join(QUEUE_FILE),
            history_file: work_dir.join(HISTORY_FILE),
            inner: Mutex::new(QueueInner {
                queue: HashMap::new(),
                history: HashMap::new(),
            }),
        };
        svc.load_queue();
        svc.load_history();
        let (q, h) = {
            let g = svc.inner.lock().unwrap();
            (g.queue.len(), g.history.len())
        };
        info!("GoogleIndexQueueService initialized — queue: {q}, history URLs: {h}");
        svc
    }

    pub fn add_to_queue(&self, url: &str, action: &str, priority: i32) -> bool {
        let mut g = self.inner.lock().unwrap();
        if g.queue.contains_key(url) {
            return false;
        }
        if Self::is_recently_submitted_inner(&g.history, url, DEDUP_DAYS) {
            return false;
        }
        g.queue.insert(url.to_string(), UrlSubmissionQueue::new(url, action, priority));
        drop(g);
        self.save_queue();
        true
    }

    pub fn add_batch_to_queue(&self, urls: &[String], action: &str, priority: i32) -> HashMap<String, bool> {
        urls.iter()
            .map(|u| (u.clone(), self.add_to_queue(u, action, priority)))
            .collect()
    }

    pub fn get_next_batch(&self, batch_size: usize) -> Vec<UrlSubmissionQueue> {
        let g = self.inner.lock().unwrap();
        let mut items: Vec<_> = g
            .queue
            .values()
            .filter(|i| i.status == "PENDING" || i.status == "FAILED")
            .filter(|i| i.retry_count < MAX_RETRY_COUNT)
            .cloned()
            .collect();
        items.sort_by(|a, b| b.effective_priority().partial_cmp(&a.effective_priority()).unwrap());
        items.truncate(batch_size);
        items
    }

    pub fn mark_as_processing(&self, url: &str) {
        let mut g = self.inner.lock().unwrap();
        if let Some(item) = g.queue.get_mut(url) {
            item.status = "PROCESSING".into();
            item.last_attempt_at = now_ms();
        }
        drop(g);
        self.save_queue();
    }

    pub fn mark_as_completed(&self, url: &str, success: bool, response: &str) {
        let mut g = self.inner.lock().unwrap();
        let Some(mut item) = g.queue.remove(url) else {
            return;
        };

        if success {
            let mut hist = UrlSubmissionHistory::new(&item.url, &item.action, true);
            hist.response = Some(response.to_string());
            g.history.entry(item.url.clone()).or_default().push(hist);
            Self::trim_history(&mut g.history, &item.url);
            drop(g);
            self.save_queue();
            self.save_history();
            return;
        }

        item.retry_count += 1;
        item.last_error = Some(response.to_string());
        if item.retry_count >= MAX_RETRY_COUNT {
            let mut hist = UrlSubmissionHistory::new(&item.url, &item.action, false);
            hist.response = Some(response.to_string());
            g.history.entry(item.url.clone()).or_default().push(hist);
            Self::trim_history(&mut g.history, &item.url);
        } else {
            item.status = "PENDING".into();
            g.queue.insert(url.to_string(), item);
        }
        drop(g);
        self.save_queue();
        self.save_history();
    }

    pub fn remove_from_queue(&self, url: &str) -> bool {
        let removed = self.inner.lock().unwrap().queue.remove(url).is_some();
        if removed {
            self.save_queue();
        }
        removed
    }

    pub fn get_queue_info(&self) -> Map<String, Value> {
        let g = self.inner.lock().unwrap();
        let pending = g.queue.values().filter(|i| i.status == "PENDING").count();
        let processing = g.queue.values().filter(|i| i.status == "PROCESSING").count();
        let failed = g.queue.values().filter(|i| i.status == "FAILED").count();
        let mut m = Map::new();
        m.insert("total".into(), json!(g.queue.len()));
        m.insert("pending".into(), json!(pending));
        m.insert("processing".into(), json!(processing));
        m.insert("failed".into(), json!(failed));
        m.insert("history_urls".into(), json!(g.history.len()));
        m
    }

    pub fn get_queue_items(&self, page: usize, page_size: usize) -> Vec<UrlSubmissionQueue> {
        let g = self.inner.lock().unwrap();
        let mut items: Vec<_> = g.queue.values().cloned().collect();
        items.sort_by(|a, b| b.effective_priority().partial_cmp(&a.effective_priority()).unwrap());
        items.into_iter().skip(page * page_size).take(page_size).collect()
    }

    pub fn get_history(&self, url: &str) -> Vec<UrlSubmissionHistory> {
        self.inner
            .lock()
            .unwrap()
            .history
            .get(url)
            .cloned()
            .unwrap_or_default()
    }

    pub fn get_recent_history(&self, limit: usize) -> Vec<UrlSubmissionHistory> {
        let g = self.inner.lock().unwrap();
        let mut all: Vec<_> = g.history.values().flatten().cloned().collect();
        all.sort_by(|a, b| b.submitted_at.cmp(&a.submitted_at));
        all.truncate(limit);
        all
    }

    fn is_recently_submitted_inner(history: &HashMap<String, Vec<UrlSubmissionHistory>>, url: &str, days: i64) -> bool {
        history
            .get(url)
            .map(|h| h.iter().any(|e| e.success && e.is_within_days(days)))
            .unwrap_or(false)
    }

    fn trim_history(history: &mut HashMap<String, Vec<UrlSubmissionHistory>>, url: &str) {
        if let Some(list) = history.get_mut(url) {
            if list.len() > MAX_HISTORY_ENTRIES_PER_URL {
                list.sort_by(|a, b| b.submitted_at.cmp(&a.submitted_at));
                list.truncate(MAX_HISTORY_ENTRIES_PER_URL);
            }
        }
    }

    fn load_queue(&self) {
        let Ok(text) = fs::read_to_string(&self.queue_file) else {
            return;
        };
        if text.trim().is_empty() {
            return;
        }
        if let Ok(list) = serde_json::from_str::<Vec<UrlSubmissionQueue>>(&text) {
            let mut g = self.inner.lock().unwrap();
            g.queue.clear();
            for item in list {
                g.queue.insert(item.url.clone(), item);
            }
        }
    }

    fn load_history(&self) {
        let Ok(text) = fs::read_to_string(&self.history_file) else {
            return;
        };
        if text.trim().is_empty() {
            return;
        }
        if let Ok(map) = serde_json::from_str::<HashMap<String, Vec<UrlSubmissionHistory>>>(&text) {
            self.inner.lock().unwrap().history = map;
        }
    }

    fn save_queue(&self) {
        let g = self.inner.lock().unwrap();
        let list: Vec<_> = g.queue.values().cloned().collect();
        drop(g);
        if let Ok(json) = serde_json::to_string_pretty(&list) {
            if let Some(parent) = self.queue_file.parent() {
                let _ = fs::create_dir_all(parent);
            }
            if let Err(e) = fs::write(&self.queue_file, json) {
                warn!("Failed to save google index queue: {e}");
            }
        }
    }

    fn save_history(&self) {
        let g = self.inner.lock().unwrap();
        let map = g.history.clone();
        drop(g);
        if let Ok(json) = serde_json::to_string_pretty(&map) {
            if let Some(parent) = self.history_file.parent() {
                let _ = fs::create_dir_all(parent);
            }
            if let Err(e) = fs::write(&self.history_file, json) {
                warn!("Failed to save google index history: {e}");
            }
        }
    }
}

File: backend-go/internal/services/google_index_queue.go
package services

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"time"
)

const (
	googleQueueFile   = "google-index-queue.json"
	googleHistoryFile = "google-index-history.json"
	maxRetryCount     = 3
	dedupDays         = int64(30)
	maxHistoryPerURL  = 10
)

type URLSubmissionQueue struct {
	URL           string `json:"url"`
	Action        string `json:"action"`
	Priority      int32  `json:"priority"`
	QueuedAt      int64  `json:"queuedAt"`
	RetryCount    int32  `json:"retryCount"`
	LastError     string `json:"lastError,omitempty"`
	Status        string `json:"status"`
	LastAttemptAt int64  `json:"lastAttemptAt"`
}

type URLSubmissionHistory struct {
	URL          string `json:"url"`
	Action       string `json:"action"`
	SubmittedAt  int64  `json:"submittedAt"`
	SubmittedDate string `json:"submittedDate"`
	Success      bool   `json:"success"`
	Response     string `json:"response,omitempty"`
	QuotaUsed    int32  `json:"quotaUsed"`
	SubmissionID string `json:"submissionId"`
}

type GoogleIndexQueueService struct {
	queueFile   string
	historyFile string
	mu          sync.Mutex
	queue       map[string]URLSubmissionQueue
	history     map[string][]URLSubmissionHistory
}

func NewGoogleIndexQueueService(workDir string) *GoogleIndexQueueService {
	_ = os.MkdirAll(workDir, 0o755)
	svc := &GoogleIndexQueueService{
		queueFile:   filepath.Join(workDir, googleQueueFile),
		historyFile: filepath.Join(workDir, googleHistoryFile),
		queue:       make(map[string]URLSubmissionQueue),
		history:     make(map[string][]URLSubmissionHistory),
	}
	svc.loadQueue()
	svc.loadHistory()
	return svc
}

func (q *URLSubmissionQueue) effectivePriority() float64 {
	ageHours := float64(nowMs()-q.QueuedAt) / (1000 * 60 * 60)
	return float64(11-q.Priority)*100 + ageHours
}

func newQueueItem(url, action string, priority int32) URLSubmissionQueue {
	if priority < 1 {
		priority = 1
	}
	if priority > 10 {
		priority = 10
	}
	return URLSubmissionQueue{
		URL:      url,
		Action:   action,
		Priority: priority,
		QueuedAt: nowMs(),
		Status:   "PENDING",
	}
}

func (s *GoogleIndexQueueService) AddToQueue(url, action string, priority int32) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.queue[url]; ok {
		return false
	}
	if s.isRecentlySubmittedLocked(url, dedupDays) {
		return false
	}
	s.queue[url] = newQueueItem(url, action, priority)
	s.saveQueueLocked()
	return true
}

func (s *GoogleIndexQueueService) AddBatchToQueue(urls []string, action string, priority int32) map[string]bool {
	out := make(map[string]bool, len(urls))
	for _, u := range urls {
		out[u] = s.AddToQueue(u, action, priority)
	}
	return out
}

func (s *GoogleIndexQueueService) GetNextBatch(batchSize int) []URLSubmissionQueue {
	s.mu.Lock()
	defer s.mu.Unlock()
	items := make([]URLSubmissionQueue, 0, len(s.queue))
	for _, item := range s.queue {
		if item.Status != "PENDING" && item.Status != "FAILED" {
			continue
		}
		if item.RetryCount >= maxRetryCount {
			continue
		}
		items = append(items, item)
	}
	sort.Slice(items, func(i, j int) bool {
		return items[i].effectivePriority() > items[j].effectivePriority()
	})
	if len(items) > batchSize {
		items = items[:batchSize]
	}
	return items
}

func (s *GoogleIndexQueueService) MarkAsProcessing(url string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if item, ok := s.queue[url]; ok {
		item.Status = "PROCESSING"
		item.LastAttemptAt = nowMs()
		s.queue[url] = item
		s.saveQueueLocked()
	}
}

func (s *GoogleIndexQueueService) MarkAsCompleted(url string, success bool, response string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	item, ok := s.queue[url]
	if !ok {
		return
	}
	delete(s.queue, url)
	if success {
		hist := newHistory(item.URL, item.Action, true)
		hist.Response = response
		s.history[item.URL] = append(s.history[item.URL], hist)
		s.trimHistoryLocked(item.URL)
		s.saveQueueLocked()
		s.saveHistoryLocked()
		return
	}
	item.RetryCount++
	item.LastError = response
	if item.RetryCount >= maxRetryCount {
		hist := newHistory(item.URL, item.Action, false)
		hist.Response = response
		s.history[item.URL] = append(s.history[item.URL], hist)
		s.trimHistoryLocked(item.URL)
	} else {
		item.Status = "PENDING"
		s.queue[url] = item
	}
	s.saveQueueLocked()
	s.saveHistoryLocked()
}

func (s *GoogleIndexQueueService) RemoveFromQueue(url string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.queue[url]; !ok {
		return false
	}
	delete(s.queue, url)
	s.saveQueueLocked()
	return true
}

func (s *GoogleIndexQueueService) GetQueueInfo() map[string]any {
	s.mu.Lock()
	defer s.mu.Unlock()
	pending, processing, failed := 0, 0, 0
	for _, item := range s.queue {
		switch item.Status {
		case "PENDING":
			pending++
		case "PROCESSING":
			processing++
		case "FAILED":
			failed++
		}
	}
	return map[string]any{
		"total":        len(s.queue),
		"pending":      pending,
		"processing":   processing,
		"failed":       failed,
		"history_urls": len(s.history),
	}
}

func (s *GoogleIndexQueueService) GetQueueItems(page, pageSize int) []URLSubmissionQueue {
	s.mu.Lock()
	defer s.mu.Unlock()
	items := make([]URLSubmissionQueue, 0, len(s.queue))
	for _, item := range s.queue {
		items = append(items, item)
	}
	sort.Slice(items, func(i, j int) bool {
		return items[i].effectivePriority() > items[j].effectivePriority()
	})
	start := page * pageSize
	if start >= len(items) {
		return nil
	}
	end := start + pageSize
	if end > len(items) {
		end = len(items)
	}
	return items[start:end]
}

func (s *GoogleIndexQueueService) GetHistory(url string) []URLSubmissionHistory {
	s.mu.Lock()
	defer s.mu.Unlock()
	return append([]URLSubmissionHistory(nil), s.history[url]...)
}

func (s *GoogleIndexQueueService) GetRecentHistory(limit int) []URLSubmissionHistory {
	s.mu.Lock()
	defer s.mu.Unlock()
	all := make([]URLSubmissionHistory, 0)
	for _, list := range s.history {
		all = append(all, list...)
	}
	sort.Slice(all, func(i, j int) bool {
		return all[i].SubmittedAt > all[j].SubmittedAt
	})
	if len(all) > limit {
		all = all[:limit]
	}
	return all
}

func (s *GoogleIndexQueueService) isRecentlySubmittedLocked(url string, days int64) bool {
	list := s.history[url]
	cutoff := nowMs() - days*24*60*60*1000
	for _, h := range list {
		if h.Success && h.SubmittedAt >= cutoff {
			return true
		}
	}
	return false
}

func (s *GoogleIndexQueueService) trimHistoryLocked(url string) {
	list := s.history[url]
	if len(list) <= maxHistoryPerURL {
		return
	}
	sort.Slice(list, func(i, j int) bool {
		return list[i].SubmittedAt > list[j].SubmittedAt
	})
	s.history[url] = list[:maxHistoryPerURL]
}

func (s *GoogleIndexQueueService) loadQueue() {
	data, err := os.ReadFile(s.queueFile)
	if err != nil || len(data) == 0 {
		return
	}
	var list []URLSubmissionQueue
	if json.Unmarshal(data, &list) != nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.queue = make(map[string]URLSubmissionQueue, len(list))
	for _, item := range list {
		s.queue[item.URL] = item
	}
}

func (s *GoogleIndexQueueService) loadHistory() {
	data, err := os.ReadFile(s.historyFile)
	if err != nil || len(data) == 0 {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	_ = json.Unmarshal(data, &s.history)
}

func (s *GoogleIndexQueueService) saveQueueLocked() {
	list := make([]URLSubmissionQueue, 0, len(s.queue))
	for _, item := range s.queue {
		list = append(list, item)
	}
	data, err := json.MarshalIndent(list, "", "  ")
	if err != nil {
		return
	}
	_ = os.WriteFile(s.queueFile, data, 0o644)
}

func (s *GoogleIndexQueueService) saveHistoryLocked() {
	data, err := json.MarshalIndent(s.history, "", "  ")
	if err != nil {
		return
	}
	_ = os.WriteFile(s.historyFile, data, 0o644)
}

func newHistory(url, action string, success bool) URLSubmissionHistory {
	ts := nowMs()
	return URLSubmissionHistory{
		URL:           url,
		Action:        action,
		SubmittedAt:   ts,
		SubmittedDate: time.Now().Format("2006-01-02"),
		Success:       success,
		SubmissionID:  genSubmissionID(),
	}
}

func genSubmissionID() string {
	return fmtSubmissionID(nowMs())
}

func fmtSubmissionID(ts int64) string {
	return fmt.Sprintf("%d-%d", ts, time.Now().UnixNano())
}

func nowMs() int64 {
	return time.Now().UnixMilli()
}
